File: Java/22.java
class Solution {
    public List<String> recurse(int n, String current, List<String> lst){
        if(n==0){
            if(!lst.contains(current)) lst.add(current);
            return lst;
        }
        for(int i = 0; i <= current.length()/2; i++) lst = recurse(n-1,current.substring(0,i) + "()" + current.substring(i),lst);
        return lst;
    }
    public List<String> generateParenthesis(int n) {
        List<String> lst = new ArrayList<>();
        lst = recurse(n,"",lst);
        return lst;
    }
}

File: Python/22.py
class Solution(object):
    def generateParenthesis(self, n):
        lst = recurse(n-1,"()",[])
        return lst
def recurse(n,current,lst):
    if n==0:
        if current not in lst:
            lst.append(current)
        return lst
    for i in range(len(current)+1):
        lst = recurse(n-1,current[:i] + "()" + current[i:],lst)
    return lst